lightgbm/src/main/scala/com/microsoft/ml/spark/lightgbm/dataset/DatasetAggregator.scala

```diff
@@ -0,0 +1,488 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark.lightgbm.dataset
+
+import com.microsoft.ml.spark.lightgbm.ColumnParams
+import com.microsoft.ml.spark.lightgbm.swig.{BaseSwigArray, ChunkedArray, DoubleChunkedArray, DoubleSwigArray,
+  FloatChunkedArray, FloatSwigArray, IntChunkedArray, IntSwigArray}
+import org.apache.spark.sql.Row
+
+import scala.collection.mutable.ListBuffer
+
+object ChunkedArrayUtils {
+  def copyChunkedArray[T](chunkedArray: ChunkedArray[T],
+                          mainArray: BaseSwigArray[T],
+                          threadRowStartIndex: Long,
+                          chunkSize: Long): Unit = {
+    // Copy in parallel on each thread
+    // First copy full chunks
+    val chunkCount = chunkedArray.getChunksCount() - 1
+    for (chunk <- 0L until chunkCount) {
+      for (in_chunk_idx <- 0L until chunkSize) {
+        mainArray.setItem(threadRowStartIndex + chunk * chunkSize + in_chunk_idx,
+          chunkedArray.getItem(chunk, in_chunk_idx, (-1).asInstanceOf[T]))
```

Comment:
Instead of doing type conversion, consider adding a context bound to T:
```
def copyChunkedArray[T: Numeric](chunkedArray: ChunkedArray[T],
...
  val num = implicitly[Numeric[T]]
...
       chunkedArray.getItem(chunk, in_chunk_idx, num.fromInt(-1)))
```

The benefits are:
1. This effectively adds a constraint to type T, such that only numeric types can be specified.
2. num.fromInt(-1) will be type safe.

Reply:
great idea, done!